src/QFRSimulator.cpp

```diff
@@ -1,6 +1,30 @@
 #include "QFRSimulator.hpp"
 
-void QFRSimulator::Simulate() {
+std::map<std::string, unsigned int> QFRSimulator::Simulate(unsigned int shots) {
+    bool has_nonunitary = false;
+    for (auto& op : *qc) {
+        if (op->isNonUnitaryOperation()) {
+            has_nonunitary = true;
+            break;
+        }
+    }
+
+    if(!has_nonunitary) {
+        single_shot();
+        return MeasureAllNonCollapsing(shots);
+    }
+
+    std::map<std::string, unsigned int> m_counter;
+
+    for (unsigned int i=0; i < shots; i++) {
```

Comment:
I suppose the simulation of the circuit is repeated for each shot due to the possibility of intermediate measurements.
In case all the measurements are at the end, this is highly inefficient. 
Any `.qasm` file that ends with `measure q->c;` should in principle be simulated once and then measured according to the requested number of shots.
Please correct me if I am wrong.

Reply:
Yeah, it's a bit more inefficient than it needs to be, but we'll address this when we switch from measuring all qubits to only measuring the ones that are actually measured in the circuit.